### PR TITLE
repo-updater: fallback onto users repos endpoint when org fails

### PR DESCRIPTION
**Pull request by kzh**
This addresses #4764. `org:<org name>` in `repositoryQuery` is special cased to hit the org repos API directly. This is to avoid reaching the 1000 search results limit. However, `org:<username>` is also valid input. This PR tries the user repos API (`users/:user/repos`) if the org repos (`orgs/:org/repos`) endpoint fails.

Side: `user:<username>` should also be special cased to hit this users endpoint in a followup PR.

### CODE REVIEW

File: pkg/extsvc/github/repos.go [hunk 1]
```diff
@@ -461,38 +453,35 @@ func (c *Client) ListPublicRepositories(ctx context.Context, sinceRepoID int64)
...
-	if err := c.requestGet(ctx, "", path, &restRepos); err != nil {
-		return nil, false, 1, err
-	}
-	repos = make([]*Repository, 0, len(restRepos))
-	for _, restRepo := range restRepos {
-		repos = append(repos, convertRestRepo(restRepo))
+	repos, err = c.listRepositories(ctx, path)
+	if err == nil {
+		// 🚨 SECURITY: must forward token here to ensure caching by token
+		c.addRepositoriesToCache("", repos)
```
Comment (keegancsmith) [inline pkg/extsvc/github/repos.go]: @tsenart @beyang Is this still a thing? I don't see any call paths to ListAffiliatedRepositories from a auth code path?

File: cmd/repo-updater/repos/github.go [hunk 2]
```diff
@@ -285,9 +285,15 @@ func (s *GithubSource) paginate(ctx context.Context, results chan *githubResult,
 // listOrg handles the `org` config option.
 // It returns all the repositories belonging to the given organization
 // by hitting the /orgs/:org/repos endpoint.
-func (s *GithubSource) listOrg(ctx context.Context, org string, results chan *githubResult) {
+//
+// It returns an error if the request fails on the first page.
+func (s *GithubSource) listOrg(ctx context.Context, org string, results chan *githubResult) (fail error) {
 	s.paginate(ctx, results, func(page int) (repos []*github.Repository, hasNext bool, cost int, err error) {
 		defer func() {
+			if err != nil && page == 1 {
```
[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/repo-updater/repos/github.go]: Agree here. It would be great to have an explicit check here, even if it means checking whether the error message contains "is not a user" or something like that.

File: cmd/repo-updater/repos/github.go [hunk 3]
```diff
@@ -300,6 +306,33 @@ func (s *GithubSource) listOrg(ctx context.Context, org string, results chan *gi
...
+// listUser returns all the repositories belonging to the given user
+// by hitting the /users/:user/repos endpoint.
+//
+// It returns an error if the request fails on the first page.
+func (s *GithubSource) listUser(ctx context.Context, user string, results chan *githubResult) (fail error) {
+	s.paginate(ctx, results, func(page int) (repos []*github.Repository, hasNext bool, cost int, err error) {
+		defer func() {
+			if err != nil && page == 1 {
+				fail, err = err, nil
+			}
```
Comment (mrnugget) [inline cmd/repo-updater/repos/github.go]: Just a thought, haven't looked deeply into it, but I wonder whether we can't refactor `s.paginate` so that we don't have to do this `defer` dance here, which I think is a little bit "bending over backwards" :) Maybe it would be enough if, instead of using `defer` we explicitly check what `s.client.ListUserRepositories` returns?

(This is no "request for changes", just an idea in case you want to try to refactor the code)

Comment (kzh) [reply cmd/repo-updater/repos/github.go]: Ooo, I'll look into this in a follow up PR for refactoring. The reason I placed this here was to catch the error before going into the `results` channel in `paginate` so I could handle it later somewhere. I do agree the defer makes the logic hard to follow. I'll put more thought into the refactor, but thanks for the suggestion!

**Comment by kzh** [comment]
Thanks for the feedback @keegancsmith @mrnugget 😁

File: cmd/repo-updater/repos/github.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kzh) [reply cmd/repo-updater/repos/github.go]: I switched it to check for a 404.

File: cmd/repo-updater/repos/github.go [hunk 4]
```diff
@@ -494,8 +527,17 @@ func (s *GithubSource) listRepositoryQuery(ctx context.Context, query string, re
 	// to directly use GitHub's org repo
 	// list API instead of the limited
 	// search API.
+	//
+	// If the org repo list API fails, we
+	// try the user repo list API.
 	if org := matchOrg(query); org != "" {
-		s.listOrg(ctx, org, results)
+		if err := s.listOrg(ctx, org, results); err != nil {
+			if uerr := s.listUser(ctx, org, results); uerr != nil {
```
[the remark this answers is not in the document]
Comment (kzh) [reply cmd/repo-updater/repos/github.go]: Gotcha!
